File: backend/src/modules/rag/ingest/parsePdf.js
import { execFile } from "child_process";
import path from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const SCRIPT_PATH = path.join(__dirname, "parse_pdf.py");

/**
 * Parses PDF file using PyMuPDF (fitz) via python child process.
 * Extracts page text and page number.
 * Returns array of pages: [{ pageNumber: 1, text: "..." }]
 *
 * // TODO: OCR support if needed in future for scanned image PDFs
 */
export function parsePdf(pdfPath) {
  return new Promise((resolve, reject) => {
    execFile(
      "python",
      [SCRIPT_PATH, pdfPath],
      { maxBuffer: 50 * 1024 * 1024, encoding: "utf-8" },
      (error, stdout, stderr) => {
        if (error) {
          return reject(new Error(`Failed to parse PDF via PyMuPDF: ${error.message} - ${stderr}`));
        }

        try {
          const pages = JSON.parse(stdout.trim());
          if (pages.error) {
            return reject(new Error(`PyMuPDF extraction error: ${pages.error}`));
          }
          resolve(pages);
        } catch (e) {
          reject(new Error(`Invalid JSON output from parse_pdf.py: ${e.message}`));
        }
      }
    );
  });
}

File: backend/src/modules/rag/ingest/parse_pdf.py
import sys
import json
import pymupdf as fitz  # PyMuPDF

if hasattr(sys.stdout, 'reconfigure'):
    sys.stdout.reconfigure(encoding='utf-8')

def extract_pdf_pages(pdf_path):
    # TODO: OCR support if needed in future for scanned image PDFs
    doc = fitz.open(pdf_path)
    pages = []
    for page_num in range(len(doc)):
        page = doc[page_num]
        text = page.get_text("text") or ""
        pages.append({
            "pageNumber": page_num + 1,
            "text": text
        })
    doc.close()
    return pages

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(json.dumps({"error": "No PDF path provided"}))
        sys.exit(1)
    
    pdf_path = sys.argv[1]
    try:
        pages = extract_pdf_pages(pdf_path)
        print(json.dumps(pages, ensure_ascii=False))
    except Exception as e:
        print(json.dumps({"error": str(e)}))
        sys.exit(1)
